File: m02_ros2_communication/m02_ros2_with_py/m02_ros2_with_py/int64_sub.py
# !/usr/bin/env python3

# ---------------------------- ROS2 Dependencies ------------------------------
import rclpy
from rclpy.node import Node

# ---------------------------- ROS2 Messages -----------------------------------
from std_msgs.msg import Int64

# ---------------------------- Subscriber Class Node ---------------------------
class MyInt64Subscriber(Node):
    """
    Simple subscriber that search for the /num_int64 info and display it.

    Attributes
    ---
    sub : Subscriptor object
        Object to subscribe to the /num_int64 topic and read the messages

    Methods
    ---
    num_callback(msg)
        Subscriber callback that display received integer value
    """
    def __init__(self):
        """
        Constructor that initialize node and instance subscriber
        """
        super().__init__("int64_sub")
        self.sub = self.create_subscription(Int64, 'num_int64', 
                                            self.num_callback, 10)
        self.sub
    
    def num_callback(self, msg):
        """
        Callback implementation when a message is recieved, so you can show it.

        Params
        ---
        msg : std_msgs.msgs/msg/Int64
            The message received, with the 'data' member that storage the 
            integer.
        """
        self.get_logger().info("Recieving the integer: %d" % msg.data)


# ------------------------------ Implementation -------------------------------
def main(args=None):
    """
    Simple program to recieving integers by subscribing to the /num_int64 
    topic.
    """

    # Initialize ROS Client Library for Python
    rclpy.init(args=args)

    # Instance subscriber object
    my_sub = MyInt64Subscriber()

    # Spin that acts as a loop
    rclpy.spin(my_sub)

    # Final steps before closing program
    my_sub.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: m02_ros2_communication/m02_ros2_with_cpp/src/int64_sub.cpp
/**
 * File: int64_sub.cpp
 * 
 * Description: Simple subscriber for integer (64 bits) implemented with the 
 * rclpp.
*/

// ------------------------- CPP standard Libraries --------------------------

#include <memory>  // General utilities to manage dynamic memory

// ------------------------ rclcpp headers -----------------------------------

#include "rclcpp/rclcpp.hpp"       // Proper include for using rclcpp

#include "std_msgs/msg/int64.hpp"  // Include of standard message for int64

// ------------------------ Implementations of namespaces --------------------

// Namespace for std::bind place holders
using std::placeholders::_1;


// ------------------------- Int64 Subscriber Class --------------------------

/**
 * Simple class for a subscriber of integers (64 bits) which inherits from the 
 * Node class
*/
class IntSub : public rclcpp::Node
{
  // Public interface
  public:
    /**
     * Constructor call that initialize the node name (using parent interface),
     * intance a subscriber object for integers by usin the 'num_int64' topic
     * and linking a callback for displaying the message received.
     * 
    */
    IntSub() : Node("int64_sub")
    {
      subscription_ = this->create_subscription<std_msgs::msg::Int64>(
      "num_int64", 10, std::bind(&IntSub::num_callback, this, _1));
    }

  // Private interface
  private:
    /**
     * Callback that will print the message received.
     * 
     * @param msg Pointer to the message received
    */
    void num_callback(const std_msgs::msg::Int64 & msg) const
    {
      RCLCPP_INFO(this->get_logger(), "Message received: '%ld'", msg.data);
    }

    // Private shared pointer for subscriber instance
    rclcpp::Subscription<std_msgs::msg::Int64>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  // ROS Client Library for C++ initialization
  rclcpp::init(argc, argv);

  // Spin until message is sent
  rclcpp::spin(std::make_shared<IntSub>());

  // Closing node
  rclcpp::shutdown();
  return 0;
}